### PR TITLE
Export CLA signatures

### PR DESCRIPTION
**Pull request by robbkidd**
Refactored some query logic in the Signatures scopes to allow looking up the earliest a user or organization signed a CLA.

Added rake tasks to export users/organizations based on whether their first CLA signature comes after a given date.

Example usage:

Individual contributors:
```
» be rake signers:export:icla date="2015-07-01"
id,user_id,signed_at,prefix,first_name,middle_name,last_name,suffix,email,phone,address_line_1,address_line_2,city,state,zip,country,icla_id
3,3,2015-08-01 20:22:09 UTC,,Jarod,,Fadel,,jarodfadel@considine.org,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
4,4,2015-09-04 07:44:22 UTC,,Florine,,Steuber,,florinesteuber@borer.name,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
6,6,2015-07-14 13:14:01 UTC,,Adaline,,Gulgowski,,adalinegulgowski@schusterfeil.biz,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
10,10,2015-09-14 10:44:29 UTC,,Alberto,,Hauck,,albertohauck@sawaynzemlak.biz,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
11,11,2015-09-27 18:41:35 UTC,,Tyrese,,Oberbrunner,,tyreseoberbrunner@green.biz,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
13,13,2015-07-08 18:58:34 UTC,,Rowena,,Littel,,rowenalittel@oconnellvonrueden.net,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
14,14,2015-08-13 22:18:54 UTC,,Mabelle,,McDermott,,mabellemcdermott@kingziemann.net,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
27,27,2015-09-05 19:16:19 UTC,,Roscoe,,Blanda,,roscoeblanda@farrellspinka.info,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
28,28,2015-07-04 08:48:16 UTC,,Abel,,Hackett,,abelhackett@greenfelderbashirian.net,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
36,36,2015-08-29 21:04:33 UTC,,Mercedes,,Hilpert,,mercedeshilpert@torp.org,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
38,38,2015-08-14 18:46:13 UTC,,Hipolito,,Paucek,,hipolitopaucek@markshintz.org,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
39,39,2015-07-10 12:28:38 UTC,,Amiya,,Larkin,,amiyalarkin@nader.info,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
40,40,2015-09-27 19:44:36 UTC,,Jaron,,Zboncak,,jaronzboncak@kirlin.name,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
43,43,2015-10-04 09:16:47 UTC,,Cassandre,,Hamill,,cassandrehamill@cartwrightpurdy.org,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
44,44,2015-07-07 00:39:38 UTC,,Gloria,,McCullough,,gloriamccullough@schmidt.biz,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
49,49,2015-10-02 04:44:30 UTC,,Pete,,Wilkinson,,petewilkinson@nadermarvin.net,(555) 555-5555,1 Main St.,,Whoville,NY,12345,USA,1
51,101,2015-10-06 19:35:08 UTC,,John,,Doe,,john@example.com,888-555-5555,123 Fake Street,,Burlington,Vermont,05401,United States,
```

Corporate contributor:
```
» be rake signers:export:ccla date="2015-07-01"
id,user_id,organization_id,ccla_id,signed_at,prefix,first_name,middle_name,last_name,suffix,email,phone,company,address_line_1,address_line_2,city,state,zip,country,created_at,updated_at
1,51,1,1,2015-09-17 13:28:22 UTC,,Ubaldo,,Runolfsdottir,,ubaldorunolfsdottir@rempel.biz,(555) 555-5555,Contributing Company 0,1 Main St.,,Whoville,NY,12345,USA,2015-10-06 19:35:08 UTC,2015-10-06 19:35:08 UTC
8,58,8,1,2015-08-05 18:56:14 UTC,,Melissa,,Bechtelar,,melissabechtelar@zboncak.net,(555) 555-5555,Contributing Company 7,1 Main St.,,Whoville,NY,12345,USA,2015-10-06 19:35:08 UTC,2015-10-06 19:35:08 UTC
10,60,10,1,2015-08-31 02:43:48 UTC,,Keenan,,Connelly,,keenanconnelly@okunevawolff.net,(555) 555-5555,Contributing Company 9,1 Main St.,,Whoville,NY,12345,USA,2015-10-06 19:35:08 UTC,2015-10-06 19:35:08 UTC
15,65,15,1,2015-09-24 17:55:01 UTC,,Mary,,Nolan,,marynolan@dickinson.info,(555) 555-5555,Contributing Company 14,1 Main St.,,Whoville,NY,12345,USA,2015-10-06 19:35:08 UTC,2015-10-06 19:35:08 UTC
24,74,24,1,2015-09-29 00:54:00 UTC,,Wade,,Upton,,wadeupton@smithamroberts.info,(555) 555-5555,Contributing Company 23,1 Main St.,,Whoville,NY,12345,USA,2015-10-06 19:35:08 UTC,2015-10-06 19:35:08 UTC
```

### CODE REVIEW

**Comment by robbkidd** [comment]
@nellshamrell For your perusal ...

File: app/models/concerns/exportable.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module Exportable
```
Comment (nellshamrell) [inline app/models/concerns/exportable.rb]: Would love to see a spec file for this module as well.

Comment (robbkidd) [inline app/models/concerns/exportable.rb]: I thought you might say that. Now to figure out a light way to spec the Exportable concern given as_csv's Demeter violations and assumptions. Setting up a sham AR::Base class may get tricky. I'll research.

File: spec/support/shared_examples/exportable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    it 'has header row of model\'s attributes' do
+      expect(reparsed_csv.first).to eq(exportable_thing.attribute_names)
+    end
+
+    it 'has a row for a model in a query' do
+      exportable_thing_attr_values = exportable_thing.attributes.values_at(*model.column_names)
+      exportable_thing_attr_values_stringified = exportable_thing_attr_values.map(&:to_s)
+      expect(reparsed_csv.second.map(&:to_s)).to eq(exportable_thing_attr_values_stringified)
```
Comment (robbkidd) [inline spec/support/shared_examples/exportable_spec.rb]: Ugh.

File: app/models/concerns/exportable.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  extend ActiveSupport::Concern
```
Comment (robbkidd) [inline app/models/concerns/exportable.rb]: Thoughts on whether to just make this a plain old Module and mix it in like plain old Ruby?

Comment (nellshamrell) [inline app/models/concerns/exportable.rb]: I'd support trying that out.

Comment (robbkidd) [inline app/models/concerns/exportable.rb]:
```
diff --git a/app/models/concerns/exportable.rb b/app/models/concerns/exportable.rb
index 35638c3..1551a40 100644
--- a/app/models/concerns/exportable.rb
+++ b/app/models/concerns/exportable.rb
@@ -1,7 +1,10 @@
 module Exportable
-  extend ActiveSupport::Concern
   require 'csv'

+  def self.included base
+    base.extend ClassMethods
+  end
+
   module ClassMethods
     def as_csv
       CSV.generate do |csv|
```

Only change made to make it PORO. Tests pass. I'm kinda meh on which way to take, maybe lean toward PORO.

File: spec/models/ccla_signature_spec.rb [hunk 4]
```diff
@@ -54,17 +56,42 @@
...
+        create(:ccla_signature, organization: organization, signed_at: 1.week.ago,
+                                user: repeat_signer, last_name: repeat_signer.last_name)
+      end
+      let!(:earliest_signature) { create(:ccla_signature, organization: organization, signed_at: 1.year.ago) }
+
+      it 'returns the earliest signature for each of the users' do
+        expect(CclaSignature.earliest_by_user).to match_array([earliest_signature, recent_signature, latest_signature])
+      end
+
+      it 'does not return an older signature by the same user' do
```
Comment (nellshamrell) [inline spec/models/ccla_signature_spec.rb]: Did you mean newer?

**Comment by robbkidd** [comment]
@nellshamrell OK. Fixed the spec description. Tests pass. I think we're good to go.

**Comment by nellshamrell** [comment]
:+1: Let's :shipit:

**Pull request merged**